Source/drlg_l1.cpp

```diff
@@ -2168,6 +2168,130 @@ void CryptPatternGroup7(int rndper)
 	PlaceMiniSetRandom(CryptPattern8, rndper);
 }
 
+bool PlaceCathedralMiniSets(lvl_entry entry)
+{
+	bool doneflag = false;
+
+	if (Quests[Q_PWATER].IsAvailable()) {
+		if (entry == ENTRY_MAIN) {
+			if (PlaceMiniSet(PWATERIN, 1, 1, 0, 0, true, -1) < 0)
+				doneflag = true;
+		} else {
+			if (PlaceMiniSet(PWATERIN, 1, 1, 0, 0, false, -1) < 0)
+				doneflag = true;
+			ViewPosition.y--;
+		}
+	}
+	if (Quests[Q_LTBANNER].IsAvailable()) {
+		if (entry == ENTRY_MAIN) {
+			if (PlaceMiniSet(STAIRSUP, 1, 1, 0, 0, true, -1) < 0)
+				return true;
+		} else {
+			if (PlaceMiniSet(STAIRSUP, 1, 1, 0, 0, false, -1) < 0)
+				doneflag = true;
+			if (entry == ENTRY_PREV) {
+				ViewPosition = Point { 20, 28 } + Displacement { setpc_x, setpc_y } * 2;
+			} else {
+				ViewPosition.y--;
+			}
+		}
+	} else if (entry == ENTRY_MAIN) {
+		if (Players[MyPlayerId].pOriginalCathedral) {
+			if (PlaceMiniSet(L5STAIRSUP, 1, 1, 0, 0, true, -1) < 0)
+				return true;
+			if (PlaceMiniSet(STAIRSDOWN, 1, 1, 0, 0, false, -1) < 0)
+				return true;
+		} else {
+			if (PlaceMiniSet(STAIRSUP, 1, 1, 0, 0, true, -1) < 0)
+				doneflag = true;
+			if (PlaceMiniSet(STAIRSDOWN, 1, 1, 0, 0, false, -1) < 0)
+				doneflag = true;
+		}
+	} else if (!Players[MyPlayerId].pOriginalCathedral && entry == ENTRY_PREV) {
+		if (PlaceMiniSet(STAIRSUP, 1, 1, 0, 0, false, -1) < 0)
+			doneflag = true;
+		if (PlaceMiniSet(STAIRSDOWN, 1, 1, 0, 0, true, -1) < 0)
+			doneflag = true;
+		ViewPosition.y--;
+	} else {
+		if (Players[MyPlayerId].pOriginalCathedral) {
+			if (PlaceMiniSet(L5STAIRSUP, 1, 1, 0, 0, false, -1) < 0)
+				doneflag = true;
+			else if (PlaceMiniSet(STAIRSDOWN, 1, 1, 0, 0, true, -1) < 0)
+				doneflag = true;
+			ViewPosition.y--;
+		} else {
+			if (PlaceMiniSet(STAIRSUP, 1, 1, 0, 0, false, -1) < 0)
+				doneflag = true;
+			if (PlaceMiniSet(STAIRSDOWN, 1, 1, 0, 0, false, -1) < 0)
+				doneflag = true;
+		}
+	}
+
+	return doneflag;
+}
+
+bool PlaceCryptMiniSets(lvl_entry entry)
+{
+	bool doneflag = false;
+
+	if (entry == ENTRY_MAIN) {
+		if (currlevel == 21) {
+			if (PlaceMiniSet(L5STAIRSTOWN, 1, 1, 0, 0, false, -1) < 0)
+				doneflag = true;
+			if (PlaceMiniSet(L5STAIRSDOWN, 1, 1, 0, 0, false, -1) < 0)
+				doneflag = true;
+			ViewPosition.y++;
+		} else {
+			if (PlaceMiniSet(L5STAIRSUPHF, 1, 1, 0, 0, true, -1) < 0)
+				doneflag = true;
+			if (currlevel != 24 && PlaceMiniSet(L5STAIRSDOWN, 1, 1, 0, 0, false, -1) < 0) {
+				doneflag = true;
+			}
+			ViewPosition.y++;
+		}
+	} else if (entry == ENTRY_PREV) {

```

Comment:
Previously this might actually have produced incorrect Crypt layouts for games that where converted from D1 to Hellfire (DevilutionX only). I see this as a self-bugfix despite being a possible break in comparability :shrug:

Reply:
There are also two entrances that are programmed incorrectly in the original Hellfire, one gets overwritten, the other case can never trigger, so there was never any real consequence.